File: android/src/main/kotlin/ch/waio/pro_video_editor/ProVideoEditorPlugin.kt
package ch.waio.pro_video_editor

import android.os.Handler
import android.os.Looper
import android.util.Log
import ch.waio.pro_video_editor.src.features.Metadata
import ch.waio.pro_video_editor.src.features.render.RenderVideo
import ch.waio.pro_video_editor.src.features.ThumbnailGenerator
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.*

/** ProVideoEditorPlugin */
class ProVideoEditorPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private lateinit var renderVideo: RenderVideo
    private lateinit var metadata: Metadata
    private lateinit var thumbnailGenerator: ThumbnailGenerator

    private val coroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "pro_video_editor")
        eventChannel =
            EventChannel(flutterPluginBinding.binaryMessenger, "pro_video_editor_progress")

        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })

        renderVideo = RenderVideo(flutterPluginBinding.applicationContext);
        metadata = Metadata(flutterPluginBinding.applicationContext)
        thumbnailGenerator = ThumbnailGenerator(flutterPluginBinding.applicationContext)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            "getMetadata" -> {
                val videoBytes = call.argument<ByteArray>("videoBytes")
                val extension = call.argument<String>("extension")

                if (videoBytes != null && extension != null) {
                    val meta = metadata.processVideo(videoBytes, extension)
                    result.success(meta)
                } else {
                    result.error(
                        "InvalidArgument", "Expected raw Uint8List (ByteArray/List<Int>)", null
                    )
                }
            }

            "getThumbnails" -> {
                val id = call.argument<String>("id") ?: ""
                val videoBytes = call.argument<ByteArray>("videoBytes")
                val extension = call.argument<String>("extension")
                val boxFit = call.argument<String>("boxFit")
                val outputFormat = call.argument<String>("outputFormat")
                val outputWidth = call.argument<Number>("outputWidth")?.toInt()
                val outputHeight = call.argument<Number>("outputHeight")?.toInt()
                val rawTimestamps = call.argument<List<Number>>("timestamps") ?: emptyList()
                val timestampsUs = rawTimestamps.map { it.toLong() }
                val maxOutputFrames = call.argument<Number>("maxOutputFrames")?.toInt()


                if (videoBytes == null ||
                    extension == null ||
                    boxFit == null ||
                    outputFormat == null ||
                    outputWidth == null ||
                    outputHeight == null ||
                    (timestampsUs == null && maxOutputFrames == null)
                ) {
                    result.error("INVALID_ARGUMENTS", "Missing or invalid arguments", null)
                    return
                }
                postProgress(id, 0.0)

                coroutineScope.launch {
                    try {
                        val thumbnails = thumbnailGenerator.getThumbnails(
                            videoBytes = videoBytes,
                            extension = extension,
                            outputFormat = outputFormat,
                            boxFit = boxFit,
                            outputWidth = outputWidth,
                            outputHeight = outputHeight,
                            timestampsUs = timestampsUs,
                            maxOutputFrames = maxOutputFrames,
                            onProgress = { progress -> postProgress(id, progress) },
                        )

                        withContext(Dispatchers.Main) {
                            postProgress(id, 1.0)
                            result.success(thumbnails)
                        }
                    } catch (e: Exception) {
                        withContext(Dispatchers.Main) {
                            result.error("THUMBNAIL_ERROR", e.message, null)
                        }
                    }
                }
            }

            "renderVideo" -> {
                val id = call.argument<String>("id") ?: ""
                val videoBytes = call.argument<ByteArray>("videoBytes")
                val imageBytes = call.argument<ByteArray?>("imageBytes")
                val rotateTurns = call.argument<Number>("rotateTurns")?.toInt()
                val cropWidth = call.argument<Number>("cropWidth")?.toInt()
                val cropHeight = call.argument<Number>("cropHeight")?.toInt()
                val cropX = call.argument<Number>("cropX")?.toInt()
                val cropY = call.argument<Number>("cropY")?.toInt()
                val bitrate = call.argument<Number>("bitrate")?.toInt()
                val scaleX = call.argument<Number>("scaleX")?.toFloat()
                val scaleY = call.argument<Number>("scaleY")?.toFloat()
                val blur = call.argument<Number>("blur")?.toDouble()
                val flipX = call.argument<Boolean>("flipX") ?: false
                val flipY = call.argument<Boolean>("flipY") ?: false
                val enableAudio = call.argument<Boolean>("enableAudio") ?: true
                val playbackSpeed = call.argument<Number>("playbackSpeed")?.toFloat()
                val startUs = call.argument<Number>("startTime")?.toLong()
                val endUs = call.argument<Number>("endTime")?.toLong()
                val inputFormat = call.argument<String>("inputFormat") ?: "mp4"
                val outputFormat = call.argument<String>("outputFormat") ?: "mp4"
                val colorMatrixList = call.argument<List<List<Double>>>("colorMatrixList")
                    ?: emptyList<List<Double>>()

                if (videoBytes == null) {
                    result.error(
                        "INVALID_ARGUMENTS",
                        "Missing parameters",
                        null
                    )
                    return
                }
                postProgress(id, 0.0)

                renderVideo.render(
                    videoBytes = videoBytes,
                    imageBytes = imageBytes,
                    inputFormat = inputFormat,
                    outputFormat = outputFormat,
                    rotateTurns = rotateTurns,
                    flipX = flipX,
                    flipY = flipY,
                    scaleX = scaleX,
                    scaleY = scaleY,
                    cropWidth = cropWidth,
                    cropHeight = cropHeight,
                    cropX = cropX,
                    cropY = cropY,
                    enableAudio = enableAudio,
                    playbackSpeed = playbackSpeed,
                    startUs = startUs,
                    endUs = endUs,
                    colorMatrixList = colorMatrixList,
                    blur = blur,
                    bitrate = bitrate,
                    onProgress = { progress -> postProgress(id, progress) },
                    onComplete = { resultBytes ->
                        postProgress(id, 1.0)
                        Handler(Looper.getMainLooper()).post {
                            result.success(resultBytes)
                        }
                    },
                    onError = { error ->
                        Log.e("RenderVideo", "Error rendering video: ${error.message}")
                    }
                )
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
        coroutineScope.cancel()
    }

    private fun postProgress(id: String, progress: Double) {
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(
                mapOf(
                    "id" to id,
                    "progress" to progress
                )
            )
        }
    }
}

File: macos/Classes/ProVideoEditorPlugin.swift
import FlutterMacOS
import Foundation

public class ProVideoEditorPlugin: NSObject, FlutterPlugin {
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "pro_video_editor", binaryMessenger: registrar.messenger)
        let eventChannel = FlutterEventChannel(name: "pro_video_editor_progress", binaryMessenger: registrar.messenger)

        let instance = ProVideoEditorPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    let metadata = VideoMetadata()
    let renderQueue = DispatchQueue(label: "RenderQueue")
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("macOS \(ProcessInfo.processInfo.operatingSystemVersionString)")

        case "getMetadata":
            guard let args = call.arguments as? [String: Any],
                  let videoBytes = args["videoBytes"] as? FlutterStandardTypedData,
                  let extensionStr = args["extension"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected arguments missing", details: nil))
                return
            }

            Task {
                do {
                    let meta = try await VideoMetadata.processVideo(videoData: videoBytes.data, ext: extensionStr)
                    result(meta)
                } catch {
                    result(FlutterError(code: "METADATA_ERROR", message: error.localizedDescription, details: nil))
                }
            }

        case "getThumbnails":
            guard let args = call.arguments as? [String: Any],
                  let id = args["id"] as? String,
                  let videoBytes = (args["videoBytes"] as? FlutterStandardTypedData)?.data,
                  let extensionStr = args["extension"] as? String,
                  let boxFit = args["boxFit"] as? String,
                  let outputFormat = args["outputFormat"] as? String,
                  let outputWidth = args["outputWidth"] as? Int,
                  let outputHeight = args["outputHeight"] as? Int else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing parameters", details: nil))
                return
            }

            let timestampsUs = (args["timestamps"] as? [NSNumber])?.map { $0.int64Value } ?? []
            let maxOutputFrames = args["maxOutputFrames"] as? Int

            postProgress(id: id, progress: 0.0)

            Task {
                let thumbnails = await ThumbnailGenerator.getThumbnails(
                    videoData: videoBytes,
                    extension: extensionStr,
                    outputFormat: outputFormat,
                    boxFit: boxFit,
                    outputWidth: outputWidth,
                    outputHeight: outputHeight,
                    timestampsUs: timestampsUs,
                    maxOutputFrames: maxOutputFrames,
                    onProgress: { progress in
                        self.postProgress(id: id, progress: progress)
                    }
                )
                self.postProgress(id: id, progress: 1.0)
                result(thumbnails)
            }

        case "renderVideo":
            guard let args = call.arguments as? [String: Any],
                  let id = args["id"] as? String,
                  let videoBytes = (args["videoBytes"] as? FlutterStandardTypedData)?.data else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing parameters", details: nil))
                return
            }

            let inputFormat = args["inputFormat"] as? String ?? "mp4"
            let outputFormat = args["outputFormat"] as? String ?? "mp4"
            let imageBytes = (args["imageBytes"] as? FlutterStandardTypedData)?.data
            let rotateTurns = args["rotateTurns"] as? Int
            let cropWidth = args["cropWidth"] as? Int
            let cropHeight = args["cropHeight"] as? Int
            let cropX = args["cropX"] as? Int
            let cropY = args["cropY"] as? Int
            let scaleX = (args["scaleX"] as? NSNumber)?.floatValue
            let scaleY = (args["scaleY"] as? NSNumber)?.floatValue
            let flipX = args["flipX"] as? Bool ?? false
            let flipY = args["flipY"] as? Bool ?? false
            let blur = args["blur"] as? Double
            let bitrate = args["bitrate"] as? Int
            let enableAudio = args["enableAudio"] as? Bool ?? true
            let playbackSpeed = (args["playbackSpeed"] as? NSNumber)?.floatValue
            let startUs = args["startTime"] as? Int64
            let endUs = args["endTime"] as? Int64
            let colorMatrixList = args["colorMatrixList"] as? [[Double]] ?? []

            postProgress(id: id, progress: 0.0)

            RenderVideo.render(
                videoData: videoBytes,
                imageData: imageBytes,
                inputFormat: inputFormat,
                outputFormat: outputFormat,
                rotateTurns: rotateTurns,
                flipX: flipX,
                flipY: flipY,
                cropWidth: cropWidth,
                cropHeight: cropHeight,
                cropX: cropX,
                cropY: cropY,
                scaleX: scaleX,
                scaleY: scaleY,
                bitrate: bitrate,
                enableAudio: enableAudio,
                playbackSpeed: playbackSpeed,
                startUs: startUs,
                endUs: endUs,
                colorMatrixList: colorMatrixList,
                blur: blur,
                onProgress: { progress in
                    self.postProgress(id: id, progress: progress)
                },
                onComplete: { outputData in
                    self.postProgress(id: id, progress: 1.0)
                    result(outputData)
                },
                onError: { error in
                    result(FlutterError(code: "RENDER_ERROR", message: error.localizedDescription, details: nil))
                }
            )

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func postProgress(id: String, progress: Double) {
        DispatchQueue.main.async {
            self.eventSink?([
                "id": id,
                "progress": progress
            ])
        }
    }
}

extension ProVideoEditorPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
